client/src/main/java/io/pravega/client/netty/impl/SessionHandler.java

```diff
@@ -0,0 +1,208 @@
+
+    /**
+     * Create a session on existing connection.
+     * @param session Session.
+     * @param rp ReplyProcessor for the specified session.
+     * @return Client Connection object.
+     */
+    public ClientConnection createSession(final Session session, final ReplyProcessor rp) {
+        Exceptions.checkNotClosed(closed.get(), this);
+        log.info("Creating Session: {} for Endpoint: {}. The current Channel is {}.", session.getSessionId(), connectionName,
+                 channel.get());
+        if (sessionIdReplyProcessorMap.put(session.getSessionId(), rp) != null) {
+            throw new IllegalArgumentException("Multiple sessions cannot be created with the same Session id {}" + session.getSessionId());
+        }
+        return new ClientConnectionImpl(connectionName, session.getSessionId(), batchSizeTracker, this);
+    }
+
+    /**
+     * Close a session. This is invoked when the ClientConnection is closed.
+     * @param clientConnection Client Connection.
+     */
+    public void closeSession(ClientConnection clientConnection) {
+        int session = ((ClientConnectionImpl) clientConnection).getSession();
+        log.info("Closing Session: {} for Endpoint: {}", session, ((ClientConnectionImpl) clientConnection).getConnectionName());
+        sessionIdReplyProcessorMap.remove(session);
+    }
+
+    /**
+     * Set the Recent Message flag. This is used to avoid sending redundant KeepAlives over the connection.
+     */
+    void setRecentMessage() {
+        recentMessage.set(true);
+    }
+
+    /**
+     * Fetch the netty channel. If {@link Channel} is null then throw a ConnectionFailedException.
+     * @return  The current {@link Channel}
+     * @throws ConnectionFailedException Throw if connection is not established.
+     */
+    Channel getChannel() throws ConnectionFailedException {
+        Channel ch = channel.get();
+        if (ch == null) {
+            throw new ConnectionFailedException("Connection to " + connectionName + " is not established.");
+        }
+        return ch;
+    }
+
+    /**
+     * This function completes the input future when the channel is registered.
+     *
+     * @param future CompletableFuture which will be completed once the channel is registered.
+     */
+    void completeWhenRegistered(final CompletableFuture<Void> future) {
+        Preconditions.checkNotNull(future, "future");
+        registeredFutureLatch.register(future);
+    }
+
+    @Override
+    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+        super.channelRegistered(ctx);
+        Channel ch = ctx.channel();
+        channel.set(ch);
+        registeredFutureLatch.release(null); //release all futures waiting for channel registration to complete.
+        log.info("Connection established with Endpoint: {} on ChannelId: {}.", connectionName, ch);
+        ch.write(new WireCommands.Hello(WireCommands.WIRE_VERSION, WireCommands.OLDEST_COMPATIBLE_VERSION), ch.voidPromise());
```

Comment:
Seems like this should come before the other setting the channel.

Reply:
Updated the code so that all the connection futures (returned by `ConnectionPool#establishConnection`) are completed once `WireCommands.Hello`is sent over the wire.